### PR TITLE
Use a cache on the backend itself to store the context.

### LINKED ISSUE #208: Improve context manager performance
As per the discussion in https://github.com/Quansight-Labs/uarray/issues/206#issuecomment-535482750, `with ua.set_backend(be)` (and all other context managers) has a cost of ~400 ns.

This could be improved by e.g. having a global cache of `SetBackendContext` objects. Such a cache would need to use weak references to avoid leaking backend instances.

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #208

### CODE REVIEW

**Comment by hameerabbasi** [comment]
This is still WIP, needs `tp_visit` on both context managers.

### Benchmarks
#### Case 1
```python
with ua.set_backend(Backend):
    pass
```

#### Case 2
```python
with ctx:
    pass
```

#### `master`
Case 1: 409 ns ± 7.46 ns
Case 2: 92.5 ns ± 1.02 ns

#### `ctx-cache`
Case 1: 258 ns ± 3.27 ns
Case 2: 88.4 ns ± 1.74 ns

**Comment by hameerabbasi** [comment]
This is now ready for review.

File: uarray/_uarray_dispatch.cxx [hunk 1]
```diff
@@ -278,10 +278,10 @@ PyObject * clear_backends(PyObject * /* self */, PyObject * args)
 template <typename T>
 class context_helper
 {
-  T new_backend_;
-  std::vector<T> * backends_;
   std::vector<size_t> sizes_;
 public:
+  T new_backend_;
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: Can you at least make this read-only, i.e. add a member function `const T & get_backend() const;`.

File: uarray/_uarray_dispatch.cxx [hunk 2]
```diff
@@ -278,10 +278,10 @@ PyObject * clear_backends(PyObject * /* self */, PyObject * args)
...
+  std::vector<T> * backends_;
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: There's no reason to make this public.

**Pull request merged**